### PR TITLE
[CMake] Add AVX for Linux & macos

### PR DESCRIPTION
**Pull request by ueqri**
#4693 , add AVX or AVX2 for GCC & Clang to extend these options to Linux & macos platform.

### CODE REVIEW

File: cmake/pcl_find_avx.cmake
Comment (kunaltyagi) [inline cmake/pcl_find_avx.cmake:12]: Why set them even for the check_cxx_source_runs?

Comment (ueqri) [reply cmake/pcl_find_avx.cmake:12]: `CMAKE_REQUIRED_FLAGS` was needed to tell the CMake which flags to use for test code compiling.

And `-march=native -mtune=native` will cause the compiler to generate the most optimizing code that matches the processor if possible.(e.g. if AVX supported, then compiler will use it)

I made `check_cxx_source_runs` test compiling with these options, hence the AVX/AVX2 code could compile and run well in those supported CPU platforms. And if the test code can't be compiled or run correctly with such optimizing flags, AVX or AVX2 was definitely unsupported.

Comment (kunaltyagi) [reply cmake/pcl_find_avx.cmake:12]: Thanks for the explanation. In that case, we might need to cache the CMAKE_REQUIRED_FLAGS in a variable and then restore from there (instead of clearing CMAKE_REQUIRED_FLAGS)

Comment (ueqri) [reply cmake/pcl_find_avx.cmake:12]: Thank you for your advice!

I think CMAKE_REQUIRED_FLAGS is such a "volatile" variable that it needs to be set to particular value (sometimes this flag leaves blank) before different functions calls including `check_cxx_source_runs`, `check_c_source_runs`, etc. And the flag variable is only for these compile or run checking fields, and won't be used for other ways (I only find the CMAKE_REQUIRED_* in those functions in official docs).

I would prefer to use set/unset method for CMAKE_REQUIRED_FLAGS.

**Comment by larshg** [comment]
I just tried to use master with a downstream project (it compiles, but makes runtime error in eigen free) and to remedy it, we need to add the AVX/SSE flags in the Config as well. Can you do this?

Both the definitions and Compile options? 

https://github.com/PointCloudLibrary/pcl/blob/085075af1830eadf0d1dd51dc87e652c4274de6f/PCLConfig.cmake.in#L427

**Comment by ueqri** [comment]
Certainly, I'll do this right now. :grinning:

**Comment by ueqri** [comment]
> Both the definitions and Compile options?

`-mavx` `-mavx2` define `__AVX(2)__`, so I only added compile options in my last commit. I didn't find a official documentation about this. However, here is some helpful information: https://stackoverflow.com/a/28939692, and I've tested gcc & clang for the predefined symbols when added `-mavx*` for compile.

**Review by larshg — APPROVED** [review]

**Comment by mvieth** [comment]
Could you please also change `SSE_FLAGS_STR` to `SSE_FLAGS` in CMakeLists.txt line 196? I think that was missed in one of the last PRs.

**Comment by ueqri** [comment]
Sure, I'll do it at once. Thanks for your advice! :grinning:

**Comment by larshg** [comment]
@SunBlack Could you have a quick look?

**Comment by SunBlack** [comment]
Looks okay from my side. I would maybe just remove the line regarding `AVX_DEFINITIONS` as I short believed we have a variable anywhere with this name,

**Comment by ueqri** [comment]
> Looks okay from my side. I would maybe just remove the line regarding `AVX_DEFINITIONS` as I short believed we have a variable anywhere with this name,

Thanks for your kind reviews and advice! :grinning: I'll fix that.

**Review by larshg — APPROVED** [review]

**Review by kunaltyagi — APPROVED** [review]
`pcl_find_sse` might also benefit from some simplifications (MSVC might not need SSE_DEFINITIONS)

**Comment by mvieth** [comment]
Please don't merge this yet, I would like to have a look at the `march=native` flag

**Comment by ueqri** [comment]
> `pcl_find_sse` might also benefit from some simplifications (MSVC might not need SSE_DEFINITIONS)

It seems to work, as I referred to OpenCV CMake of finding SSE. It doesn't set definitions for compiler like `-D__SSE(2)__` and assumes the options `/arch:SSE(2)` will define them then.

But I haven't find an official docs with explicit explanation about it. These two most-related docs just mention the `_M_IX86_FP` (defined when SSE option set) & `__AVX*__` macros setting by MSVC.

So I don’t know whether the simplifications will work fine for SSE build.

**Comment by ueqri** [comment]
> Please don't merge this yet, I would like to have a look at the `march=native` flag

Thanks for your reviews! I'd be pleased to fine tune the PR if needed! :grinning:

**Comment by SunBlack** [comment]
@ueqri You could try it on godbolt with most common compilers to check it. The only documentation related the SSE macros I fond is [for the Intel compiler](https://software.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/macros/additional-predefined-macros.html)

**Comment by ueqri** [comment]
> @ueqri You could try it on godbolt with most common compilers to check it. The only documentation related the SSE macros I fond is [for the Intel compiler](https://software.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/macros/additional-predefined-macros.html)

Thanks for your advice and the documentation is quite helpful! I've tried godbolt, it's such an awesome and powerful tool to compile and test online :grinning:, but for MSVC, there's a flaw that it can compile but not execute the code. So I check it in two ways:

#### First method
I use the [test compile code](https://github.com/PointCloudLibrary/pcl/blob/4f69ae448b832887fe9acb8ea90aff5dacecbf6e/cmake/pcl_find_sse.cmake#L60-L76) in CMake, and choose `x64/x86-msvc latest(19.28)/oldest(19.14)` one by one to test compile. To my surprise, **without any provided flag**(e.g. `/arch:SSE` or `/arch:SSE2`), the compile is done all right. And I change it to AVX test compile code to find out the weird behavior, it's all the same. I am really confused about it.

Here's an example of the test: https://godbolt.org/z/zj7e8z7MT

#### Second method
I access a Windows platform with VS2019(16.9.4) to build and run the following snippet:
<details>
 <summary>Code snippet to test SSE predefined macros</summary>

```c++
#include <iostream>
int main() {
#ifdef __SSE4_2__
    std::cout << "SSE4.2 macro exists" << std::endl;
#endif
#ifdef __SSE4_1__
    std::cout << "SSE4.1 macro exists" << std::endl;
#endif
#ifdef __SSSE3__
    std::cout << "SSE3.1 macro exists" << std::endl;
#endif
#ifdef __SSE3__
    std::cout << "SSE3   macro exists" << std::endl;
#endif
#ifdef __SSE2__
    std::cout << "SSE2   macro exists" << std::endl;
#endif
#ifdef __SSE__
    std::cout << "SSE    macro exists" << std::endl;
#endif
#ifdef __AVX2__
    std::cout << "AVX2   macro exists" << std::endl;
#endif
#ifdef __AVX__
    std::cout << "AVX    macro exists" << std::endl;
#endif
    return 0;
}
```

</details>

I open SSE/SSE2 option `Project > Properties > Configuration Properties > C/C++ > Code Generation > Enable Enhanced Instruction Set` for all `Configuration` & `Platform` as below:

<details>
 <summary>Details of opening SSE/SSE2 option</summary>

Screenshot_2021-04-28_02-30-52

</details>

But no SSE definition exists (both x86 and x64 build). And I even meet this warning `command line warning D9002: ignoring unknown option '/arch:SSE2'` in x64 build (this doesn't occur in x86 build). I try to opt the AVX2 as comparison, then `__AVX2__` & `__AVX__` are highlighted and stdout shows AVX2 macros are defined. 

<details>
 <summary>Open AVX2 options as comparison</summary>

Screenshot_2021-04-28_02-40-10
Screenshot_2021-04-28_02-43-09

</details>

Here are some useful pages I've found about the situation: 
https://stackoverflow.com/a/53139592
https://github.com/weidai11/cryptopp/issues/445
https://bugzilla.mozilla.org/show_bug.cgi?id=1245693

***

And it seems that, there is no need to specify the SSE(2) in x64 build at least after VS2017, and the SSE(2) macros is defined as default in Windows according to my test and the [Intel documentation](https://software.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/macros/additional-predefined-macros.html) below:
> __SSE__ : On Windows, defined as '1'. It is undefined when option /arch:IA32 is specified. 
__SSE2__ : On Windows, defined as '1' by default or when option /arch:SSE2, /QxSSE2, /QaxSSE2, or higher processor targeting options are specified. 

(Although it's for Intel C++ compiler, the options are quite similar to MSVC)

**Comment by SunBlack** [comment]
> And it seems that, there is no need to specify the SSE(2) in x64 build at least after VS2017, and the SSE(2) macros is defined as default in Windows according to my test and the [Intel documentation](https://software.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/macros/additional-predefined-macros.html) below:

Well support of SSE2 is mandatory on all AMD64 platforms ;-). Question is now: When `PCL_ENABLE_SSE` is not set, should be disable it? (`/arch:IA32`) Maybe it would be better to provide an Enum instead of a checkbox: `Compiler default`, `Off`, `On` (first option maybe just for MSVC or did you tried GCC/Clang as well?)

**Comment by SunBlack** [comment]
I remember me, that we had a few weeks ago trouble in our company as someone wanted to use our software on a CPU, which doesn't provide the same CPU feature set than the CPU we used to compile your code. So maybe we should in generally switch to a combobox so it is possible to select the feature set (without using a VM with a old virtual CPU ;-) )

**Comment by ueqri** [comment]
@SunBlack Thanks for your comments! :blush:

> did you tried GCC/Clang as well?

I tested the snippet with SSE(2) macros check, with both GCC and Clang, in both Linux and Windows. It seems that, whether SSE(2) is set by default in x86 compile depends on the platform(Linux or Windows).
- In Linux, SSE(2) is not default in x86 build, both for GCC and Clang.
- In Windows, SSE(2) is always set by default no matter it's compiled for x86 or x64.(a bit weird feature)
<details>
 <summary>Details of the test</summary>

#### In Linux

6-linux


#### In Windows

I switch to LLVM for VS project build.
Screenshot_2021-04-29_01-44-50

Clang 32-bit build, the SSE(2) macros are defined by default.
2-x86

Clang 64-bit build, the SSE(2) macros are undoubtedly defined.
3-x64

TDM-GCC-4.9.2 32-bit build, the SSE(2) macros are still defined by default.
4-gcc32

TDM-GCC-4.9.2 32-bit build, the SSE(2) macros are also defined.
5-gcc64

</details>

> So maybe we should in generally switch to a combobox so it is possible to select the feature set (without using a VM with a old virtual CPU ;-) )

Sounds a good idea, should we use multiple-option in CMake to implement it :thinking: ? I'll be glad to contribute for it :grinning:

**Comment by SunBlack** [comment]
> Sounds a good idea, should we use multiple-option in CMake to implement it 🤔 ? I'll be glad to contribute for it 😀

No, use a [combo box](https://blog.kitware.com/constraining-values-with-comboboxes-in-cmake-cmake-gui/) for it.

Another question: In the property window `/arch:AVX` and `/arch:SSE` are within the same combo box - so it is not possible to disable when AVX is enabled? If yes: How we respect this in CMake?

**Comment by ueqri** [comment]
Given to [AVX-SSE transition penalties](https://software.intel.com/content/www/us/en/develop/articles/avoiding-avx-sse-transition-penalties.html), we'd better use only one of these instruction sets and implement this restriction in CMake(by multiple options check for CLI and combo box for GUI). I'd like to find the method to disable SSE in different platforms. If it exists, things will be easy to done.

**Comment by larshg** [comment]
> Please don't merge this yet, I would like to have a look at the `march=native` flag

FYI on vcpkg https://github.com/microsoft/vcpkg/pull/17156

**Comment by kunaltyagi** [comment]
What's actionable on this PR? Seems like we can't do the same for SSE. Is it just the review by @mvieth ?

Based on @larshg 's comment, do we want to add a super-option: `DISABLE_VECTORIZATION` or something? If no, cool (less logic in CMake 😄 ). If yes, this PR or another PR?

**Comment by larshg** [comment]
There is already `PCL_ENABLE_AVX `and `PCL_ENABLE_SSE ` - which is also the one that is used in the PR in VCPKG to disable it.

However, maybe there should be a specific `PCL_ENABLE/ALLOW_MARCHNATIVE` ? and otherwise it defaults to the "standard" SSE options?

**Comment by larshg** [comment]
As for the SSE defines, there are more possible defines than there are /arch options in visual studio. So for windows I would say the definitions are still required.

AVX and AVX2 is possible to set only by /arch.

**Comment by kunaltyagi** [comment]
Thanks for the clarification. Let's move the discussion about avx/sse combo option else where. And wait for the a review on this PR :)

**Comment by larshg** [comment]
Also, these defines doesn't seem to be used, so we can remove them:
https://github.com/PointCloudLibrary/pcl/blob/b19c99a5f45950c0e519e631d09adb0c5257596e/pcl_config.h.in#L54-L59

**Comment by ueqri** [comment]
> Is it not necessary to add ${AVX_FLAGS} everywhere after ${SSE_FLAGS} in CMakeLists.txt? Currently that is just the case for MSVC, but not for GCC or CLANG

Thanks for your reminder @mvieth ! I've added the AVX_FLAGS for GCC and Clang compile flags.

> CMake currently does not use the march=native flag, but says that it does ("Using CPU native flags for SSE optimization"). I think the problem is that in pcl_find_sse.cmake, set(SSE_FLAGS ...) is used instead of something like string(APPEND SSE_FLAGS ...), overwriting the previous contents of SSE_FLAGS. It would be great if you could fix that

Sorry for still not sure about the solution. Could we just remove the following codes and add a specific PCL_ENABLE/ALLOW_MARCHNATIVE for `march=native` flag:

https://github.com/PointCloudLibrary/pcl/blob/b19c99a5f45950c0e519e631d09adb0c5257596e/cmake/pcl_find_sse.cmake#L7-L19

> Also, these defines doesn't seem to be used, so we can remove them

Thanks for your advice @larshg ! I've removed them. :grinning:

**Comment by ueqri** [comment]
> @ueqri I have opened a PR against your PR, to move the -march=native part appending it after setting the SSE_FLAGS.

Sorry for not really understanding the means. I've found recent PR are about QT updating, not very relevant to march or SSE_FLAGS. 

Could you please give me more hints or suggestions about it or this PR?

**Comment by larshg** [comment]
> > @ueqri I have opened a PR against your PR, to move the -march=native part appending it after setting the SSE_FLAGS.
> 
> Sorry for not really understanding the meanings. I've found recent PR are about QT updating, not very relevant to `-march` or SSE_FLAGS.
> 
> Could you please give me more hints or suggestions about it or this PR?

Here at your fork :) https://github.com/ueqri/pcl/pulls

**Comment by ueqri** [comment]
Oh yes! :blush: Thanks for your kind reminders!

**Comment by larshg** [comment]
I think we should add PCL_ALLOW_MARCHNATIVE as an option like here:
https://github.com/PointCloudLibrary/pcl/blob/master/cmake/pcl_options.cmake#L43-L45

And wrap this condition in it:
https://github.com/PointCloudLibrary/pcl/pull/4698/files#diff-73719a14e929c85350713acd2132c1eaad561bab49f11f86bc21942bd517fa11R221-R228

And let it be off by default. Then people can enable it, if they want to compile for optimal performance on their specific CPU on linux.
However, I assume that SSE and AVX are more common and they can be enabled by default.

Should they all be off by default and enabled by the user? @kunaltyagi @mvieth

**Comment by kunaltyagi** [comment]
Enable by default in release mode. Disabled otherwise. This gives the users speed for release and readability for debug. If they wish for portable release binaries, they can opt-in.

The distribution managers usually provide their own custom set of flags so we can ignore that use-case (since they can manually enable/disable it)

**Comment by SunBlack** [comment]
> Enable by default in release mode. Disabled otherwise. This gives the users speed for release and readability for debug. If they wish for portable release binaries, they can opt-in.

I don't like it when there are different flags used in release and debug mode, as it makes debugging difficult when different code is executed. So it can happen that an error occurs in release mode that is not reproducible in debug mode and therefore harder to find because you might not know that there are differences.

**Comment by larshg** [comment]
> > Enable by default in release mode. Disabled otherwise. This gives the users speed for release and readability for debug. If they wish for portable release binaries, they can opt-in.
> 
> I don't like it when there are different flags used in release and debug mode, as it makes debugging difficult when different code is executed. So it can happen that an error occurs in release mode that is not reproducible in debug mode and therefore harder to find because you might not know that there are differences.

There are many (optimization)flags that are different when compiling Debug vs Release. I think you would have even more trouble debugging with release flags, as all additional help would be optimized away?

**Comment by SunBlack** [comment]
Of course there are optimization flags, so e.g. assertions will be not executed. But usually it is still executing the same code.

**Comment by kunaltyagi** [comment]
Would you want to enable AVX and SSE in debug builds as well? (But not march=native)

And add march=native in Release.

How does this sound?

**Comment by SunBlack** [comment]
I would add a CMake option where the user can decide what they want - so I wouldn't differentiate between release and debug as the user can configure it themselves.

**Comment by kunaltyagi** [comment]
Defaults matter. Specially when most people just perform a `cmake .. && make install -j2`

Options are mostly for power users or for those performing an atypical installation (cross-machine, cross-compilation, etc.)

In my opinion, if AVX/SSE capability is present, we should enable it. It is no different than detection OpenMP and enabling it.

If it turns out, we're not propagating flags properly, or something similar, we can always roll back to the current more conservative flags.

**Comment by larshg** [comment]
I have added the PCL_ENABLE_MARCHNATIVE, so it can be enabled/disabled.

**Comment by ueqri** [comment]
That's totally OK, definitely! :grinning: And I was wondering if there is something I could work on before this PR being merged.

File: cmake/pcl_find_sse.cmake
Comment (kunaltyagi) [inline cmake/pcl_find_sse.cmake:221]: Can we remove this and only depend on the condition with HAVE_MARCH?

File: cmake/pcl_find_sse.cmake
Comment (kunaltyagi) [inline cmake/pcl_find_sse.cmake:223]: march=native is not a SSE flag. Thoughts on AVX enabled, but SSE disabled configuration? Can we prevent the user from shooting themselves in the foot?

IMO, march=native is separate from AVX and SSE configurations (AVX and SSE are kinda-standard opsets, but march native can also use non-standard ops)

**Review by kunaltyagi — COMMENTED** [review]
Rest LGTM

**Comment by kunaltyagi** [comment]
@ueqri Any updates on this? (I know GSoC is going on, just that it's been almost a month). Would like to release 1.12.1 with this.

File: cmake/pcl_find_sse.cmake
Comment (mvieth) [reply cmake/pcl_find_sse.cmake:221]: Seems to be possible, but I would say it is better (for safety and for understanding) to check everything: The flags should only be added if the user wants them (`PCL_ENABLE_MARCHNATIVE`), if we are not crosscompiling (`NOT CMAKE_CROSSCOMPILING`), _and_ if the compiler knows the flag (`HAVE_MARCH`)

File: cmake/pcl_find_sse.cmake
Comment (mvieth) [reply cmake/pcl_find_sse.cmake]: A cleaner solution might be to unify pcl_find_sse.cmake and pcl_find_avx.cmake, and have a single variable (e.g. `ARCH_FLAGS`) that contains sse, avx, and march flags, instead of `SSE_FLAGS` and `AVX_FLAGS`. But I would say the current state of this PR still works, and in the interest of getting this merged soon I would propose to not do any major changes here.
I think the cmake options `PCL_ENABLE_SSE`, `PCL_ENABLE_AVX`, and `PCL_ENABLE_MARCHNATIVE` give good control over which flags are added.

File: cmake/pcl_find_sse.cmake
Comment (kunaltyagi) [reply cmake/pcl_find_sse.cmake:221]: L11 ensures that HAVE_MARCH is enabled only if this exact condition is valid.

Comment (mvieth) [reply cmake/pcl_find_sse.cmake:221]: But when would we set the `-mtune=native` flag? Only checking that `HAVE_MARCH==false` does not seem to be sufficient since `CMAKE_CROSSCOMPILING` could be true or `PCL_ENABLE_MARCHNATIVE` could be false/off (I would interpret it so that if `PCL_ENABLE_MARCHNATIVE` is false/off, neither `march=native` nor `mtune=native` is set)

**Review by kunaltyagi — APPROVED** [review]

**Review by mvieth — APPROVED** [review]

**Review by larshg — APPROVED** [review]

**Pull request merged**
